backend/unit-tests/focs-2-3-tests.ml

```diff
@@ -0,0 +1,148 @@
+
+let testNoRepeats = 
+	let rec findDupes = function
+	|[] -> false
+	|x::xs -> List. exists ((=)x) xs || findDupes xs in
+	let rndNum1 = (Random.int 10000) + 1 in
+	let rndNum2 = (Random.int rndNum1) + 1 in
+	let rndList1 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum1 ))) @ (intsTo (Random.int rndNum1 )) in
+	let rndList2 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum2 ))) @ (intsTo (Random.int rndNum2 )) in
+	assert_equal false (findDupes (Submission.subtract rndList1 rndList2));;
+
+let testSubmissionOrder = 
+	let rec checkOrder xs ys = match xs, ys with
+	|[], [] -> true
+	|[], ys -> true
+	|xs, [] -> false
+	|x::xs, y::ys -> if (x==y) then checkOrder xs ys
+			 else checkOrder (x::xs) ys in
+	let rndNum1 = (Random.int 10000) + 1 in
+	let rndNum2 = (Random.int rndNum1) + 1 in
+	let rndList1 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum1 ))) @ (intsTo (Random.int rndNum1 )) in
+	let rndList2 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum2 ))) @ (intsTo (Random.int rndNum2 )) in
+	assert_equal true (checkOrder (Submission.subtract rndList1 rndList 2) rndList1);;
+
+
+(*union*)
+
+let testOnEmptyLists = assert_equal [] (Submission.union [] []);;
+
+let testOnOneEmptyList = 
+	let rndList = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int 10000))) @ (intsTo (Random.int 10000)) in
+	assert_equal rndList (Submission.union [] rndList);;
+
+let testOnNonEmptyLists = 
+	let rndNum1 = (Random.int 10000) + 1 in
+	let rndNum2 = (Random.int rndNum1) + 1 in
+	let rndList1 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum1 ))) @ (intsTo (Random.int rndNum1 )) in
+	let rndList2 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum2 ))) @ (intsTo (Random.int rndNum2 )) in
+	assert_equal rndList1 (Submission.subtract rndList1 rndList2);;
+
+let testNoRepeats = 
+	let rec findDupes = function
+	|[] -> false
+	|x::xs -> List. exists ((=)x) xs || findDupes xs in
+	assert_equal false (findDupes (Submission.union rndList (10001::rndList)));;
+
+let testSubmissionOrder = 
+	let rec checkOrder xs ys = match xs, ys with
+	|[], [] -> true
+	|[], ys -> true
+	|xs, [] -> false
+	|x::xs, y::ys -> if (x==y) then checkOrder xs ys
+			 else checkOrder (x::xs) ys in
+	let rndNum1 = (Random.int 10000) + 1 in
+	let rndNum2 = (Random.int rndNum1) + 1 in
+	let rndList1 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum1 ))) @ (intsTo (Random.int rndNum1 )) in
+	let rndList2 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum2 ))) @ (intsTo (Random.int rndNum2 )) in
+	assert_equal true (checkOrder (Submission.subtract rndList1 rndList 2) rndList1);;
+
+(*exf*)
+
+let testOnEmptyList = assert_equal [] (Submission.exf (fun x-> x+1) []);;
+
+let testOnExample = assert_equal ([2;8] || [8;2]) (Submission.exf (fun x-> x+1) [9;3;2;2;8]);;
```

Comment:
Can we test exf on some large, autogenerated lists please? Can use permutations function to make testing the output order easier.

Reply:
Also, can we put exf in a separate file, please?

Reply:
Would it be FoCS supervision 2 in one file and supervision 3 in another then?

Reply:
I think one file per question would be best. Currently there's one file for each of the exercises listed at http://svwork.ireland.cx/tasks.php and it would be good if the new tests could follow a similar convention.